src/main/java/io/confluent/connect/elasticsearch/ElasticsearchClient.java

```diff
@@ -362,23 +364,28 @@ public void afterBulk(long executionId, BulkRequest request, Throwable failure)
    * @return the return value of the called function
    */
   private <T> T callWithRetries(String description, Callable<T> function) {
-    try {
-      return RetryUtil.callWithRetries(
-          description,
-          function,
-          config.maxRetries(),
-          config.retryBackoffMs()
-      );
-    } catch (Exception e) {
-      throw new ConnectException("Failed to " + description + ".", e);
-    }
+    return RetryUtil.callWithRetries(
+        description,
+        function,
+        config.maxRetries() + 1,
+        config.retryBackoffMs()
+    );
   }
 
   /**
-   * Closes all of the connection and thread resources of the client.
+   * Closes all the connection and thread resources of the client.
    */
-  private void closeConnections() {
-    executorService.shutdown();
+  private void closeResources() {
+    bulkExecutorService.shutdown();
+    try {
+      if (!bulkExecutorService.awaitTermination(config.flushTimeoutMs(), TimeUnit.MILLISECONDS)) {
+        bulkExecutorService.shutdownNow();
+      }
+    } catch (InterruptedException e) {
+      bulkExecutorService.shutdownNow();
+      Thread.currentThread().interrupt();
+      log.warn("Interrupted while awaiting for executor service shutdown.", e);
+    }
```

Comment:
The `config.flushTimeoutMs()` could be set to a large value, and if so that could prevent interrupting executor threads that are blocked in backoffs before a retry.

Instead, I'd suggest using a constant duration (e.g., 5 seconds) that is not configurable. This will prevent users from accidentally configuring the connector one way and then having it becomes really slow to terminate the tasks.